linkerd/protocol/base-http/src/main/scala/io/buoyant/linkerd/http/ZipkinTracePropagator.scala

```diff
@@ -0,0 +1,94 @@
+package io.buoyant.linkerd.protocol.http
+
+import com.twitter.finagle.{SimpleFilter, Stack}
+import com.twitter.finagle.buoyant.Sampler
+import com.twitter.finagle.http.util.StringUtil
+import com.twitter.finagle.tracing.{Flags, SpanId, TraceId}
+import com.twitter.util.Try
+import io.buoyant.linkerd.{TracePropagator, TracePropagatorInitializer}
+import io.buoyant.router.http.{HeadersLike, RequestLike}
+
+class ZipkinTracePropagator[Req, H: HeadersLike](implicit requestLike: RequestLike[Req, H]) extends TracePropagator[Req] {
+  /**
+   * Read the trace id from the request, if it has one.
+   */
+  override def traceId(req: Req): Option[TraceId] = {
+    var traceId = ZipkinTrace.get(requestLike.headers(req))
+
```

Comment:
Shouldn't we be clearing the trace headers here also?  As in `LinkerdHeaders.Ctx.Trace.clear(req.headers)`?

Reply:
I removed that line of code, it didn't make sense to me to clear l5d-ctx header in a Zipkin propagator plugin. Am I missing something?

Reply:
The way that trace propagation works in Linkerd is that the trace context is read from the incoming request (in a server filter), a child span is created, and the whole context is stored in a thread local.  When the request is ready to be sent out (in a client filter), the trace context is read from the thread local and written into the request headers.

This means that when we read the trace context from a request, we should clear those headers.

Reply:
Here's my idea of fixing this: clear the old tracing context just before setting the new tracing context. At least for Zipkin propagation, we cannot clear tracing context during get traceid because we need it to be available for get sampler call (which will use x-b3-sampled and x-b3-flags headers). Also, we cannot clear tracing context during get sampler call because we need it to be available for the get traceid call (same headers). Get traceid and get sampler do not depend on each other, so the context must be preserved for both calls. Note that this is only specific for Zipkin propagation. I've pushed the latest changes implementing this idea. Please let me know if it's OK!

Reply:
After looking at this code some more and reading https://github.com/openzipkin/b3-propagation again, I think this can actually be simpler.  The `X-B3-Sampled` only supports values of `0` and `1` which means that the value is fully recoverable from the `traceId` in the `traceId.sampled` field.

Therefore my suggestion is to have `get` read the traceId and clear all the headers.  Then, the `sampler` method can simply ignore the `req` entirely and use `Trace.id.sampled` to create a sampler.  

This relies on the fact that `traceId` will be called before `sampler` and will set the traceId into the thread local, but looking at `TraceInitializer`, I think this is a fair assumption.  (sorry if that contradicts some of my earlier suggestions)  We should also add some documentation to the TracePropagator trait to make these assumptions explicit.

Reply:
Makes sense, here's how I'm thinking to implement your suggestion:
- have a default implementation of 
object TraceInitializer {
  abstract class ServerFilter {
    def sampler(req: Req): Option[Sampler] = {
      - if Trace.id set return Sampler(Trace.id.sampled) else None
- ZipkinTracePropagator will not overwrite the default sampler implementation above.
- LinkerdTracePropagator will have to overwrite it, no actual change in its source code.

Should be an easy change but I want to do testing before posting it. Let me know if we're on the same page with this. Thanks.

Reply:
I don't think we need to make that the default implementation of `sampler`.  Safer to just explicitly make it the implementation for ZipkinTracePropagator.